src/services/services.ts

```diff
@@ -6804,6 +6794,67 @@ namespace ts {
             return [];
         }
 
+        /**
+         * Checks if position points to a valid position to add JSDoc comments, and if so,
+         * returns the appropriate template. Otherwise returns an empty string.
+         * Valid positions are
+         * * outside of comments, statements, and expressions, and
+         * * preceding a function declaration.
+         * 
+         * Hosts should ideally check that:
+         * * The line is all whitespace up to 'position' before performing the insertion.
+         * * If the keystroke sequence "/\*\*" induced the call, we also check that the next
+         * non-whitespace character is '*', which (approximately) indicates whether we added 
+         * the second '*' to complete an existing (JSDoc) comment.
+         * @param fileName The file in which to perform the check.
+         * @param position The (character-indexed) position in the file where the check should
+         * be performed.
+         */
+        function getDocCommentTemplateAtPosition(fileName: string, position: number): TextInsertion {
+            let start = new Date().getTime();
+            let sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+            log("getDocCommentTemplateAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));

```

Comment:
This seems excessive.  Is there a reason we're logging this?  If this is an important metric, maybe we should add the logging for this inside of getCurrentSourceFile itself.